### PR TITLE
KAFKA-3622: Use descriptive error message if port number is missing from url

### PR DESCRIPTION
**Pull request by peterableda**
A new exception is thrown in _parseAndValidateAddresses_ method when the port number is missing from the url. 

The change was not trivial as the previous implementation of _getHost_ and _getPort_ both returned _null_ if ex. the port number was missing from the url.
To fix this behaviour I added some extra test cases and examples to the unit tests with the expected behaviour and modified the _HOST_PORT_PATTERN_ regex and the above mentioned methods.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 1]
```diff
@@ -50,7 +50,7 @@
 
     // This matches URIs of formats: host:port and protocol:\\host:port
     // IPv6 is supported with [ip] pattern
-    private static final Pattern HOST_PORT_PATTERN = Pattern.compile(".*?\\[?([0-9a-zA-Z\\-.:]*)\\]?:([0-9]+)");
+    private static final Pattern HOST_PORT_PATTERN = Pattern.compile(".*?(?:(?:\\[([0-9a-zA-Z:]*)\\])|([0-9a-zA-Z\\-.]*))(?::([0-9]+))?");
```
Comment (peterableda) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: .*?  **<- pattern to match the protocol part of the url**
    (?:(?:\[([0-9a-zA-Z:]*)\])  **<- match the host as an IPv6 address**
              |([0-9a-zA-Z\-.]*))  **<- match the host as a non-IPv6 address (doesn't contain ':')**
(?::([0-9]+))? **<-match the "optional" port number**

File: clients/src/main/java/org/apache/kafka/clients/ClientUtils.java [hunk 2]
```diff
@@ -40,8 +40,12 @@
             if (url != null && url.length() > 0) {
                 String host = getHost(url);
                 Integer port = getPort(url);
-                if (host == null || port == null)
+                if (host == null)
                     throw new ConfigException("Invalid url in " + CommonClientConfigs.BOOTSTRAP_SERVERS_CONFIG + ": " + url);
```
Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:45]: Should this say "Missing host in..." now?

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java (hunk 1 above)
Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: Are both host groups really needed? Does the one that contains ':' handle both?

We have regex in other places in the project that do similar parsing we may want to keep in sync:
- https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/utils/Utils.java#L53
- https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/cluster/BrokerEndPoint.scala#L27
- https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/cluster/EndPoint.scala#L29

Comment (peterableda) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: The issue with having only one type was inputs like: "127.0.0.1:8000" resulted in host: "127.0.0.1:8000" port: null matches as the we allowed ':' characters in the host part. Now we only allow the ':' character if it is surrounded with [] brackets.

I will check the remaining regex statements to see what can I do there.

File: clients/src/main/java/org/apache/kafka/clients/ClientUtils.java (hunk 2 above)
Comment (peterableda) [inline clients/src/main/java/org/apache/kafka/clients/ClientUtils.java:45]: I can change it if you prefer that one. I assume though that it's pretty rare for users to forget about the host part of the url, the issue there is probably some invalid character or formatting.

**Comment by peterableda** [comment]
@granthenke I checked the other regex patterns in the EndPoint/BrokerEndPoint files and they are somewhat different. I think it's possible to create one pattern for all but not sure what was the reason we created different ones. If there is a need to have one pattern for all use-cases I'm happy to work on it but it might be better to create a separate JIRA/pull request for that.

**Comment by rehevkor5** [comment]
Don't give up now @peterableda :)

**Comment by vvcephei** [comment]
Hi @peterableda ,

It seems like this PR stalled. I'll close it out for now, but if you or anyone else want to resume this work, please feel free to re-open it (or start a new one)!

Thanks,
John

**Pull request closed without merging**